prime-router/src/main/kotlin/ReportStreamFilter.kt

```diff
@@ -2,39 +2,206 @@ package gov.cdc.prime.router
+    JsonSubTypes.Type(BundleResourceFilter::class, name = "bundleResource"),
+    JsonSubTypes.Type(BundleObservationFilter::class, name = "bundleObservation"),
+    JsonSubTypes.Type(BundleConditionFilter::class, name = "bundleCondition"),
+)
+interface BundleFilterable {
+    /**
+     * Check if a [bundle] passes this filter
+     * @return whether the bundle passed
+     */
+    fun pass(bundle: Bundle): Boolean
+}
 
+/**
+ * Interface for pruning T resources from a bundle
+ */
+/**
+ * A bundle filter that uses resources as the basis for filtering
+ */
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
 @JsonSubTypes(
-    JsonSubTypes.Type(CodeStringConditionFilter::class, name = "codeString"),
-    JsonSubTypes.Type(ConditionLookupConditionFilter::class, name = "conditionLookup"),
-    JsonSubTypes.Type(FHIRExpressionConditionFilter::class, name = "fhirExpression"),
+    JsonSubTypes.Type(ConditionCodePruner::class, name = "conditionCode"),
+    JsonSubTypes.Type(ConditionKeywordPruner::class, name = "conditionKeyword"),
+    JsonSubTypes.Type(FHIRExpressionPruner::class, name = "fhirExpression"),
 )
-abstract class ConditionFilter(val value: String) {
-    // TODO: change to .filter() or .passes() method and combine with legacy filter
-    abstract fun codes(): List<String>
+interface BundlePrunable<T> {
+    /**
+     * Check if a [resource] in a [bundle] passes this filter
+     * @return whether the observation passed
+     */
+    fun evaluateResource(bundle: Bundle, resource: T): Boolean
+
+    fun fetchResources(bundle: Bundle): List<T>
+
+    /**
+     * Check if a [bundle] passes this filter
+     * @return whether the bundle passed
+     */
+    fun prune(bundle: Bundle): List<T>
 }
 
-class CodeStringConditionFilter(value: String) : ConditionFilter(value) {
-    private val codeList = value.split(",").map { it.trim() }
-    override fun codes() = codeList
+open class BundleResourceFilter<T : Resource>(val resourceFilter: BundlePrunable<T>) : BundleFilterable {
+    override fun pass(bundle: Bundle): Boolean {
+        return resourceFilter.fetchResources(bundle).any {
+            resourceFilter.evaluateResource(bundle, it)
+        }
+    }
 }
 
-class ConditionLookupConditionFilter(value: String) : ConditionFilter(value) {
-    override fun codes() = emptyList<String>()
+class BundleObservationFilter(
+    val observationFilter: BundlePrunable<Observation>,
+) : BundleResourceFilter<Observation>(observationFilter)
+
+class BundleConditionFilter(
+    val conditionFilter: BundlePrunable<Observation>,
+) : BundleResourceFilter<Observation>(conditionFilter) {
+    override fun pass(bundle: Bundle): Boolean =
+        conditionFilter.fetchResources(bundle).filter { observation ->
+            conditionFilter.evaluateResource(bundle, observation)
+        }.let { observations ->
+            // never pass a bundle with only AOE conditions
```

Comment:
This has been reworked so that the AOE check is no longer necessary.

The business logic is that we want to include AOEs if we found a condition we're interested in. So instead of configuring them in the filter as a code that passes (which results in the edge case of a bundle with all AOEs), the filter logic adds AOEs only if it found a condition that passes.

Reply:
Nice! And if a receiver for some reason does not want any or certain AOEs, we can have a separate "pruner" for that.